File: demos/demo_Messi_DTW.cpp
#include "../commons/dataloaders.hpp"
#include "../lib/daisy.hpp"
#include <chrono>
#include <algorithm>

int main(){
    
    daisy::idx_t n_database = 200000;
    unsigned long long dim = 96;
    unsigned long long n_query = 10;
    daisy::idx_t k = 5;

    float *database = loadRandomData(n_database, dim, 100, true);
    float *query = loadRandomData(n_query, dim, 50, true);

    printf("Loaded %llu database points and %llu query points with dimension %llu\n", n_database, n_query, dim);

    daisy::Messi messi_search(daisy::DistanceType::DTW);
    messi_search.setNumThreads(4);

    int warp_window = std::max(1, static_cast<int>(dim * 0.1));
    messi_search.setWarpingWindow(warp_window);  

    messi_search.buildIndex(database, n_database, dim);

    daisy::idx_t *I = new daisy::idx_t[n_query * k];
    float *D = new float[n_query * k];
    messi_search.searchIndex(query, n_query, k, I, D);

    for (daisy::idx_t i = 0; i < n_query; i++)
    {
        printf("Query %llu: ", i);
        for (daisy::idx_t j = 0; j < k; j++)
        {
            printf("%llu ", I[i * k + j]);
        }
        printf("\n");
    }

    delete[] database;
    delete[] query;
    delete[] I;
    delete[] D; 

    return 0;
}

File: demos/demo_Messi_DTW.py
import sys
import os
import numpy as np

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from daisy import DistanceType, Messi

def main():
    
    n_database = 200000
    dim = 96
    n_query = 10
    k = 5

    np.random.seed(100)   
    db = np.random.randn(n_database, dim).astype(np.float32)

    np.random.seed(50)
    query = np.random.randn(n_query, dim).astype(np.float32)

    index = Messi(DistanceType.DTW)
    index.setWarpingWindow(int(dim * 0.1))

    index.setNumThreads(1)
    index.buildIndex(db)

    I, D = index.searchIndex(query, k)

    for query_num in range(n_query):
        print(f"Query {query_num}:")
        print("Distances:", D[query_num])
        print("Indices:", I[query_num])
        print()

if __name__ == "__main__":
    main()